tla/counterfeit-cycles/counterfeit_cycles.tla

```diff
@@ -0,0 +1,295 @@
+
+-----------------------------------------------------------------------------
+(***************************************************************************)
+(* Initialization                                                          *)
+(***************************************************************************)
+
+SubnetsInit ==
+    \* Initialize the state of subnets.
+    /\ subnets = [
+        s \in SUBNETS |-> [
+            \* The true balance of (finalized) cycles on the subnet.
+            \* In this model, subnets do not maliciously manipulate this balance.
+            \* Note that the balance can be negative if the subnet maliciously
+            \* sends more cycles than it has.
+
+            \* The balance of (finalized) cycles on the subnet.
+            \* In this model, subnets do not maliciously manipulate this balance.
+            balance |-> STARTING_BALANCE_PER_SUBNET,
+
+            \* The unfinalized cycles received from other subnets.
+            \* Cycles received from other subnets are initially unfinalized
+            \* until they are approved by the ledger.
+            unfinalized |-> [sb \in SUBNETS |-> 0],
+
+            \* A queue of outgoing messages from the subnet to the ledger.
+            msgsToLedger |-> <<>>,
+
+            \* Whether or not the subnet is honest. This isn't strictly required
+            \* in this model.
+            \* Initially, all subnets behave honestly.
+            honest |-> TRUE
+         ]
+     ]
+    /\ subnetMsgs = {}
+    /\ numDishonestActions = 0
+    /\ numTransfers = 0
+
+LedgerInit ==
+  \* Initialize the state of the ledger.
+  ledger = [
+    \* The balance of each subnet.
+    balances |-> [s \in SUBNETS |-> STARTING_BALANCE_PER_SUBNET],
+
+    \* The queue of messages to be processed.
+    msgs |-> [s \in SUBNETS |-> <<>>]
+  ]
+
+Init ==
+  /\ LedgerInit
+  /\ SubnetsInit
+
+-----------------------------------------------------------------------------
+(***************************************************************************)
+(* Type checks                                                             *)
+(***************************************************************************)
+
+TransferMessage ==
+  \* A message to transfer cycles between subnets.
+  [id: Nat, type : {TRANSFER}, from : SUBNETS, to : SUBNETS, amount : Nat \ {0}]
+
+\* The set of all possible messages that are sent/received by subnets.
+SubnetMessages ==
+  \* Transferring cycles from one subnet to another.
+  TransferMessage
+    \union
+
+  \* Approving a cycles transfer.
+  \* This message is sent by the ledger to indicate that the `amount` sent is
+  \* now finalized.
+  [id: Nat, type : {APPROVE}, from : SUBNETS, to : SUBNETS, amount : Nat \ {0}]
+
+LedgerMessages ==
+  \* The set of all possible messages that can be received by the Ledger.
+  TransferMessage
+
+SubnetsTypeOK ==
+  /\ \A s \in subnetMsgs: s \in SubnetMessages
+  /\ subnets \in [SUBNETS -> [
```

Comment:
I think the way that TLC will evaluate this is to actually construct the set on the right. Not that I understand the first thing about their codebase, but it looks like it:
https://github.com/tlaplus/tlaplus/blob/fa4968f60a0a460da4b7e182330680ce93bd96aa/tlatools/org.lamport.tlatools/src/tlc2/tool/impl/Tool.java#L2591
(there are other mentions of OPCode_In, but roughly they seem to all be evaluating the right-hand side).
So if the analysis turns out to be slow, you might want to turn off the TypeOK checks for larger models, or rewriting the checks to be more verbose (like `DOMAIN subnets = SUBNETS /\ ...`).